src/metabase/api/card.clj

```diff
@@ -619,24 +619,25 @@
              context     :question
              qp-runner   qp/process-query-and-save-execution!}}]
   {:pre [(u/maybe? sequential? parameters)]}
-  (let [run   (or run
-                  ;; param `run` can be used to control how the query is ran, e.g. if you need to
-                  ;; customize the `context` passed to the QP
-                  (^:once fn* [query info]
-                   (qp.streaming/streaming-response [context export-format]
-                     (binding [qp.perms/*card-id* card-id]
-                       (qp-runner query info context)))))
-        card  (api/read-check (Card card-id))
-        query (-> (assoc (query-for-card card parameters constraints middleware)
-                         :async? true)
-                  (update :middleware merge {:js-int-to-string?      true
-                                             :ignore-cached-results? ignore_cache}))
-        info  {:executed-by  api/*current-user-id*
-               :context      context
-               :card-id      card-id
-               :dashboard-id dashboard-id}]
+  (let [run       (or run
+                      ;; param `run` can be used to control how the query is ran, e.g. if you need to
+                      ;; customize the `context` passed to the QP
+                      (^:once fn* [query info card-name]
+                       (qp.streaming/streaming-response [context export-format (u/slugify card-name)]
+                         (binding [qp.perms/*card-id* card-id]
+                           (qp-runner query info context)))))
+        card      (api/read-check (db/select-one [Card :name :dataset_query :cache_ttl :collection_id] :id card-id))
+        card-name (:name (db/select-one [Card :name] :id card-id))
```

Comment:
this looks like an extra query where there's a `(db/select-one [Card :name ...])` on the line above

Reply:
Did you consider any alternatives? One thing that stands out to me is that you added a single string parameter to the `run` interface. There seem to be some alternatives to me that might be worth considering:
1. There is a `info` object that has `card-id` and `dashboard-id`. I'm not a fan of throwing random information into context maps, but putting the card into this map seems reasonable, and then having the streaming-response grab the name from the card there seems like a good strategy.
2. If this prefix isn't required in most codepaths, remove it from the `run` interface and just have the `streaming-response` call here use the closure by moving the binding of the card above it.
```clojure
(let [card (api/read-check (db/select-one [Card :name :dataset_query :cache_ttl :collection_id] :id card-id))
      run  (or run
               (^:once fn* [query info]
                ;; don't change the contract, just use the binding above
                (qp.streaming/streaming-response [context export-format (u/slugify (:name card))]
                    (binding [qp .perms/*card-id* card-id]
                      (qp-runner query info context)))))
      ]
  ...)
```

3. If you do like the card-name in the `run` signature, rename it perhaps to `file-prefix` or something more generic? Other callers might want to use this and indicating this is the file prefix (perhaps useful) rather than card name (not useful to anyone other than cards) would aid in understanding i think.

Reply:
Good catch about the extra query, I'd meant to remove that 🤦‍♂️

> Did you consider any alternatives?

I considered putting it into the `info` map, but I wasn't sure if that was good form here. The docstring for the `Info` schema says it "is used for informational purposes to record information about how a query was executed in QueryExecution and other places" which doesn't quite match what I'm using `card-name` for. But if it's OK, this might be the best option.

Re-ordering the bindings and using a closure is neat, but I don't think that would work for exports of public questions (in `metabase.api.public`)? It passes in a custom `run` function that needs access to card-name.

Reply:
Yeah I see. Whatever feels the most structured to you. Maybe your solution here is the best, they are all tradeoffs. Just making sure others were considered and it sounds like that was the case. 👍

Reply:
On further thought, the `info` map might be the way to go, since `card-id` is in there already. I'll try it out and see how it feels.